File: server/symbology/api/routes/groups.py
"""Company Groups API routes."""
from typing import List

from fastapi import APIRouter, HTTPException, Query, status
from symbology.api.routes.companies import _company_to_response
from symbology.api.schemas import CompanyGroupResponse, GeneratedContentResponse
from symbology.database.company_groups import (
    CompanyGroup,
    get_company_group_by_slug,
    list_company_groups,
)
from symbology.database.generated_content import get_company_group_analysis
from symbology.utils.logging import get_logger

logger = get_logger(__name__)
router = APIRouter()


def _group_to_response(group: CompanyGroup, include_companies: bool = False) -> CompanyGroupResponse:
    """Convert a CompanyGroup model to CompanyGroupResponse."""
    companies = None
    if include_companies:
        companies = [_company_to_response(c) for c in group.companies]

    # Get latest analysis summary
    latest_summary = None
    try:
        analyses = get_company_group_analysis(group.id, limit=1)
        if analyses and analyses[0].content:
            latest_summary = analyses[0].content[:500]
    except Exception as e:
        logger.warning("failed_to_get_latest_analysis", slug=group.slug, error=str(e))

    return CompanyGroupResponse(
        id=group.id,
        name=group.name,
        slug=group.slug,
        description=group.description,
        sic_codes=group.sic_codes or [],
        member_count=len(group.companies),
        created_at=group.created_at,
        updated_at=group.updated_at,
        companies=companies,
        latest_analysis_summary=latest_summary,
    )


@router.get(
    "",
    response_model=List[CompanyGroupResponse],
    status_code=status.HTTP_200_OK,
    responses={
        500: {"description": "Internal server error"},
    },
)
async def list_groups(
    skip: int = Query(0, ge=0, description="Number of groups to skip"),
    limit: int = Query(50, ge=1, le=100, description="Maximum number of groups to return"),
):
    """List all company groups."""
    logger.info("api_list_groups", skip=skip, limit=limit)

    groups = list_company_groups(limit=limit, offset=skip)
    return [_group_to_response(g) for g in groups]


@router.get(
    "/{slug}",
    response_model=CompanyGroupResponse,
    status_code=status.HTTP_200_OK,
    responses={
        404: {"description": "Group not found"},
        500: {"description": "Internal server error"},
    },
)
async def get_group_detail(slug: str):
    """Get group detail including member companies."""
    logger.info("api_get_group_detail", slug=slug)

    group = get_company_group_by_slug(slug)
    if not group:
        raise HTTPException(status_code=404, detail=f"Group not found: {slug}")

    return _group_to_response(group, include_companies=True)


@router.get(
    "/{slug}/analysis",
    response_model=List[GeneratedContentResponse],
    status_code=status.HTTP_200_OK,
    responses={
        404: {"description": "Group not found"},
        500: {"description": "Internal server error"},
    },
)
async def get_group_analysis(
    slug: str,
    limit: int = Query(5, ge=1, le=20, description="Maximum analyses to return"),
):
    """Get analyses for a company group."""
    logger.info("api_get_group_analysis", slug=slug, limit=limit)

    group = get_company_group_by_slug(slug)
    if not group:
        raise HTTPException(status_code=404, detail=f"Group not found: {slug}")

    analyses = get_company_group_analysis(group.id, limit=limit)

    return [
        GeneratedContentResponse(
            id=a.id,
            content_hash=a.content_hash,
            short_hash=a.get_short_hash(),
            company_id=a.company_id,
            company_group_id=a.company_group_id,
            description=a.description,
            document_type=a.description,
            source_type=a.source_type.value,
            created_at=a.created_at,
            total_duration=a.total_duration,
            input_tokens=a.input_tokens,
            output_tokens=a.output_tokens,
            warning=a.warning,
            content=a.content,
            summary=a.summary,
            model_config_id=a.model_config_id,
            system_prompt_id=a.system_prompt_id,
            user_prompt_id=a.user_prompt_id,
            source_document_ids=[str(d.id) for d in a.source_documents],
            source_content_ids=[str(c.id) for c in a.source_content],
        )
        for a in analyses
    ]

File: ui/src/lib/server/db/groups.ts
import { db } from '../db';
import type { Selectable } from 'kysely';
import type {
	CompanyGroupResponse,
	CompanyResponse,
	GeneratedContentResponse
} from '$lib/api-types';
import type { CompanyGroups } from './types';

export async function getCompanyGroups(limit: number = 50): Promise<CompanyGroupResponse[]> {
	const groups = await db
		.selectFrom('company_groups')
		.selectAll()
		.orderBy('name', 'asc')
		.limit(limit)
		.execute();

	return Promise.all(groups.map((g) => toGroupResponse(g, false)));
}

export async function getCompanyGroupBySlug(slug: string): Promise<CompanyGroupResponse | null> {
	const group = await db
		.selectFrom('company_groups')
		.selectAll()
		.where('slug', '=', slug)
		.executeTakeFirst();

	if (!group) return null;

	return toGroupResponse(group, true);
}

export async function getGroupFrontpageSummary(slug: string): Promise<string | null> {
	const group = await db
		.selectFrom('company_groups')
		.select('id')
		.where('slug', '=', slug)
		.executeTakeFirst();

	if (!group) return null;

	const row = await db
		.selectFrom('generated_content')
		.select('content')
		.where('company_group_id', '=', group.id)
		.where('content_stage', '=', 'company_group_frontpage')
		.orderBy('created_at', 'desc')
		.limit(1)
		.executeTakeFirst();

	return row?.content ?? null;
}

export async function getGroupAnalysis(
	slug: string,
	limit: number = 5
): Promise<GeneratedContentResponse[]> {
	const group = await db
		.selectFrom('company_groups')
		.select('id')
		.where('slug', '=', slug)
		.executeTakeFirst();

	if (!group) return [];

	const rows = await db
		.selectFrom('generated_content')
		.selectAll()
		.where('company_group_id', '=', group.id)
		.where('content_stage', '=', 'company_group_analysis')
		.orderBy('created_at', 'desc')
		.limit(limit)
		.execute();

	return Promise.all(
		rows.map(async (row) => {
			const [docIds, contentIds] = await Promise.all([
				db
					.selectFrom('generated_content_document_association')
					.select('document_id')
					.where('generated_content_id', '=', row.id)
					.execute(),
				db
					.selectFrom('generated_content_source_association')
					.select('source_content_id')
					.where('parent_content_id', '=', row.id)
					.execute()
			]);

			return {
				id: row.id,
				content_hash: row.content_hash,
				short_hash: row.content_hash?.slice(0, 12) ?? null,
				company_id: row.company_id,
				description: row.description,
				document_type: row.description,
				form_type: row.form_type,
				source_type: row.source_type,
				created_at: toISOString(row.created_at),
				total_duration: row.total_duration,
				input_tokens: row.input_tokens,
				output_tokens: row.output_tokens,
				warning: row.warning,
				content: row.content,
				summary: row.summary,
				model_config_id: row.model_config_id,
				system_prompt_id: row.system_prompt_id,
				user_prompt_id: row.user_prompt_id,
				source_document_ids: docIds.map((r) => r.document_id),
				source_content_ids: contentIds.map((r) => r.source_content_id)
			};
		})
	);
}

export async function getGroupsForCompany(companyId: string): Promise<CompanyGroupResponse[]> {
	const memberships = await db
		.selectFrom('company_group_membership')
		.select('company_group_id')
		.where('company_id', '=', companyId)
		.execute();

	if (memberships.length === 0) return [];

	const groupIds = memberships.map((m) => m.company_group_id);
	const groups = await db
		.selectFrom('company_groups')
		.selectAll()
		.where('id', 'in', groupIds)
		.execute();

	return Promise.all(groups.map((g) => toGroupResponse(g, true)));
}

async function toGroupResponse(
	group: Selectable<CompanyGroups>,
	includeCompanies: boolean
): Promise<CompanyGroupResponse> {
	// Get member count
	const members = await db
		.selectFrom('company_group_membership')
		.select('company_id')
		.where('company_group_id', '=', group.id)
		.execute();

	let companies: CompanyResponse[] | null = null;
	if (includeCompanies && members.length > 0) {
		const companyIds = members.map((m) => m.company_id);
		const companyRows = await db
			.selectFrom('companies')
			.selectAll()
			.where('id', 'in', companyIds)
			.execute();

		companies = companyRows.map((c) => ({
			id: c.id,
			name: c.name,
			display_name: c.display_name,
			ticker: c.ticker,
			exchanges: c.exchanges ?? [],
			sic: c.sic,
			sic_description: c.sic_description,
			fiscal_year_end: c.fiscal_year_end
				? new Date(c.fiscal_year_end as unknown as string).toISOString().split('T')[0]
				: null,
			former_names: (c.former_names as Array<{ name: string; date_changed: string }>) ?? [],
			summary: null
		}));
	}

	// Get latest analysis summary
	let latestAnalysisSummary: string | null = null;
	const frontpage = await db
		.selectFrom('generated_content')
		.select('content')
		.where('company_group_id', '=', group.id)
		.where('content_stage', '=', 'company_group_frontpage')
		.orderBy('created_at', 'desc')
		.limit(1)
		.executeTakeFirst();

	if (frontpage?.content) {
		latestAnalysisSummary = frontpage.content;
	} else {
		const analysis = await db
			.selectFrom('generated_content')
			.select('content')
			.where('company_group_id', '=', group.id)
			.where('content_stage', '=', 'company_group_analysis')
			.orderBy('created_at', 'desc')
			.limit(1)
			.executeTakeFirst();
		if (analysis?.content) {
			latestAnalysisSummary = analysis.content.slice(0, 500);
		}
	}

	return {
		id: group.id,
		name: group.name,
		slug: group.slug,
		description: group.description,
		sic_codes: group.sic_codes ?? [],
		member_count: members.length,
		created_at: toISOString(group.created_at),
		updated_at: toISOString(group.updated_at),
		companies,
		latest_analysis_summary: latestAnalysisSummary
	};
}

function toISOString(val: unknown): string {
	if (val instanceof Date) return val.toISOString();
	if (val == null) return '';
	return String(val);
}
